Sources/TidepoolKit/Auth/Authorization.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  Authorization.swift
+//  TidepoolKit
+//
+//  Created by Pete Schwamb on 4/15/23.
+//  Copyright © 2023 Tidepool Project. All rights reserved.
+//
+
+import Foundation
+import AppAuth
+
+
+// Testing seam for OAuth2 based authentication library, AppAuth-iOS
+
+protocol Authorization: AnyObject {
+    var currentAuthorizationFlow: OIDExternalUserAgentSession? { get set }
+
+    // Used for triggering interactive user authentication/authorization flow
+    func presentAuth(request: OIDAuthorizationRequest, presenting: UIViewController) async throws -> AuthorizationState
+
+    // Used for refreshing token
+    func requestToken(_ request: OIDTokenRequest) async throws -> AuthorizationState
+}
+
+protocol AuthorizationState {
+    var accessToken: String? { get }
+    var accessTokenExpirationDate: Date? { get }
+    var refreshToken: String? { get }
+}
+
+// Extensions for AppAuth to conform to testing seam
+
+extension OIDAuthState: AuthorizationState {
+    var accessToken: String? {
+        lastTokenResponse?.accessToken
+    }
+
+    var accessTokenExpirationDate: Date? {
+        lastTokenResponse?.accessTokenExpirationDate
+    }
+}
+
+extension OIDTokenResponse: AuthorizationState {}
+
+
+extension OIDServiceConfiguration {
+    convenience init?(from config: ProviderConfiguration) {
+        guard let authorizationEndpoint = URL(string: config.authorizationEndpoint),
+              let tokenEndpoint = URL(string: config.tokenEndpoint),
+              let issuer = URL(string: config.issuer)
+        else {
+            return nil
+        }
+
+        var endSessionEndpoint: URL?
+        if let endSessionEndpointStr = config.endSessionEndpoint {
+            endSessionEndpoint = URL(string: endSessionEndpointStr)
+        } else {
+            endSessionEndpoint = nil
+        }
+
+        self.init(
+            authorizationEndpoint: authorizationEndpoint,
+            tokenEndpoint: tokenEndpoint,
+            issuer: issuer,
+            registrationEndpoint: nil,
+            endSessionEndpoint: endSessionEndpoint)
+    }
+}
+
+class AppAuthAuthorization: Authorization {
+    var currentAuthorizationFlow: OIDExternalUserAgentSession?
+
+    func presentAuth(request: OIDAuthorizationRequest, presenting: UIViewController) async throws -> AuthorizationState {
+        // Present the authentication session using AppAuth
+        let result: (OIDAuthState, OIDExternalUserAgentSession) = try await withCheckedThrowingContinuation { continuation in
+            DispatchQueue.main.async {
```

Comment:
To stick within the world of new the concurrency model, you can mark this function as `@MainActor` or `await MainActor.run { ...`

Reply:
Is main queue even needed here?

Reply:
TBH, I really want to drop AppAuth and just use ASWebAuthenticationSession directly; we're not getting much from AppAuth at this point. But yes, we do need to call this from the main thread, I think. The docs don't specify, but it presents a view controller, and tracing through the code, I don't see it ever switching over to main inside the library for this call.